src/base1/test-protocol.js

```diff
@@ -79,4 +79,34 @@ QUnit.test("host must ensure that init is the first message", async function (as
     };
 });
 
-QUnit.start();
+QUnit.test("host must return an error when 'channel' is not given in 'open'", async function (assert) {
+    assert.expect(2);
+    let done = assert.async(2);
+
+    let connection = await connect();
+
+    // ensure that the server closes the connection on protocol error
+    connection.onclose = () => done();
+
+    connection.oncontrol = message => {
+        if (message.command === "init")
+            return;
+
+        assert.equal(message.command, "close");
+        assert.equal(message.problem, "protocol-error");
+
+        done();
+    };
+
+    connection.control({ command: "init", version: 1 });
+    connection.control({ command: "open", payload: "fsread", path: "/etc/passwd" });
+});
+
+/*
+ * Some of these tests will trigger cockpit-ws or cockpit-bridge to print out
+ * warnings (on protocol errors, for example). Let the test server know that
+ * before starting the tests, so it doesn't treat those messages as fatal.
+ */
+fetch(encodeURI("/mock/expect-warnings")).then(() => {
+    QUnit.start();
+});
```

Comment:
This now disables warning detection for *all* protocol tests. I wonder if it wouldn't  be better to fetch /expect-warnings at the beginning of this new test only, and call /unexpect-warnings at the end again? Or, more simply, have all the "strict" tests at the top, and then disable warnings, and run all  tests with warnings afterwards.

Reply:
Agreed. I toyed around with sending actual messages that are expected for every test. Alas that's not really possible, because glib's `g_test_expect_message()` tests every message after if and the bridge and web server are logging a log of other things as well. There's no way to insert the expected message at the right time.

I've changed the patch to put this test into a qunit module with before and after hooks, which let the server know that we're expecting error messages.